### ISSUE TITLE
[data grid] Implement Custom Filters

### ISSUE THREAD

**Issue by lisaWriteJava** [mui/mui-x#12978]
### The problem in depth

So I have this:  [Sandbox](https://codesandbox.io/p/sandbox/material-demo-forked-gq8pvs?file=%2Fdemo.tsx%3A30%2C25-30%2C39) which I forked. It's implmenting a not contains and it's sort of what I'd like but not really.  I'd like to:
1. not hard code the filter model like

```
const filterModel = {
  items: [{ columnField: "name", value: "Derek", operatorValue: "noContain" }],
};
```

 I want any string that the user wishes to filter to have that custom not contains operator.    

2. I'd like to know if it is possilbe to do this at a grid level instead of per column  
3. I also wishing to implement a custom filter is not
4. The last item is when I click the 'X' the filter menu doesn't disappear. I think our users would like for this to happen. It seems that outside of the custom filters this doesn't happen. 

### Your environment

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Chrome Browser

System:
OS: Linux 3.10 CentOS Linux 7 (Core)
Binaries:
Node: 16.20.0 - ~/.nvm/versions/node/v16.20.0/bin/node
npm: 7.24.2 - ~/node_modules/.bin/npm
pnpm: Not Found
Browsers:
Chrome: 120.0.6099.71
npmPackages:
@emotion/react: latest => 11.11.3
@emotion/styled: latest => 11.11.0
@mui/base: 5.0.0-beta.32
@mui/core-downloads-tracker: 5.15.5
@mui/icons-material: ^5.15.3 => 5.15.5
@mui/lab: ^5.0.0-alpha.159 => 5.0.0-alpha.161
@mui/material: ^5.15.3 => 5.15.5
@mui/private-theming: 5.15.5
@mui/styled-engine: 5.15.5
@mui/styles: ^5.15.3 => 5.15.5
@mui/system: 5.15.5
@mui/types: 7.2.13
@mui/utils: ^5.15.4 => 5.15.5
@mui/x-data-grid: 6.19.1
@mui/x-data-grid-generator: ^6.16.2 => 6.19.1
@mui/x-data-grid-premium: ^6.19.1 => 6.19.1
@mui/x-data-grid-pro: 6.19.1
@mui/x-license-pro: ^6.10.2 => 6.10.2
@types/react: 17.0.75
react: ^17.0.2 => 17.0.2
react-dom: ^17.0.2 => 17.0.2
styled-components: latest => 6.1.8
```
</details>

**Search keywords**: custom filters datagrid
**Order ID**: 66765

**Comment by michelengelen** [comment]
Hey @lisaWriteJava ... where did you fork that demo from?
The version that is being used here is only `v4` and the package has been deprecated. Check the npm repos page for it: [npm](https://www.npmjs.com/package/@material-ui/data-grid).

You can do all of this with the [filtering feature](https://mui.com/x/react-data-grid/filtering/).

**Comment by lisaWriteJava** [comment]
The example cam from stack overflow but, when I look at the documentation and the [sand box here](https://mui.com/x/react-data-grid/filtering/customization/#system-CustomRatingOperator.js)  the rating is hardcoded .  I don't see how the filter can be generalized to be used on any column or if I have multiple columns reused.  It also doesn't closed filter when the X is clicked it just clears

**Comment by lisaWriteJava** [comment]
I've alos created this: https://i34mnv--run.stackblitz.io and tried to implment the not contains that existed in the first sandbox but, it doesn't show and I don't see how I can get it to be generic to use the other string columns nor do I see how to make it be used in other columns. Please advise on what's wrong. It's not showing nor is it reusable.

**Comment by michelengelen** [comment]
OK, so to go over your questions 1by1:
1. This is possible with the [controlled filter model](https://mui.com/x/react-data-grid/filtering/#controlled-filters).
2. There is a demo on how to do that in the docs: 
3. Use a [custom operator](https://mui.com/x/react-data-grid/filtering/customization/#create-a-custom-operator) or [wrap a built-in operator](https://mui.com/x/react-data-grid/filtering/customization/#wrap-built-in-operators)
4. If you want to have a quick filter outside of the grid we have a demo on that as well: [Quick filter outside of the grid](https://mui.com/x/react-data-grid/filtering-recipes/#quick-filter-outside-of-the-grid)

---

> the rating is hardcoded 

it's only hard coded in the initial state.

>  It also doesn't closed filter when the X is clicked it just clears.

The X is not meant to close the panel, but to remove filters, or clear the last one. This is expected behavior.

---

Side note: the linked stackblitz demo does not work.

**Comment by lisaWriteJava** [comment]
Sorry about that here is the stack biz with my attempt: https://stackblitz.com/edit/react-nqs1xg?file=Demo.js

1. I've attempted to get the new filter to work with an additional column but it isn't workign. Not sure what I'm doing wrong.  
2. I've taken out the hard coding in my demo but I still don't see how I can use the 1 filter (not contains ) in multiple columns or even reuse it for the multiple tables we have. 
3.   I'm not fully understanding the custom filters to know how to achieve this.

**Comment by michelengelen** [comment]
> Sorry about that here is the stack biz with my attempt: https://stackblitz.com/edit/react-nqs1xg?file=Demo.js
> 
> 1. I've attempted to get the new filter to work with an additional column but it isn't working. Not sure what I'm doing wrong.
> 2. I've taken out the hard coding in my demo but I still don't see how I can use the 1 filter (not contains) in multiple columns or even reuse it for the multiple tables we have.
> 3. I'm not fully understanding the custom filters to know how to achieve this.

1. you need to apply the filter to every column you want it to be used in. From [this section](https://mui.com/x/react-data-grid/filtering/customization/#create-a-custom-operator):
```jsx
  const columns = React.useMemo(
    () =>
      data.columns.map((col) =>
        col.field === 'rating'
          ? {
              ...col,
              filterOperators: ratingOnlyOperators,
            }
          : col,
      ),
    [data.columns],
  );
```
2. For that you need to handle the filter state outside of the grids and apply the new filter value on every change to the input. This might be a performance bottle neck, so be careful with how many grids you apply this to.
3. Everything you need to know is described [here](https://mui.com/x/react-data-grid/filtering/customization/#customize-the-operators)

**Comment by lisaWriteJava** [comment]
I attempted to implement the not contains in the stackbiz code but I didn't
see. It show in the filter opinion. I don't understand the documentation
enough to correct it and implement properly. What did I do wrong?

We have about 15 grids and each has about 5-10:columns needing this
capability which is why I wanted to find out if there is a way to create
the functionality to be reusable.

If I can get my attempts to work I think it will set me up to get this
working.

**Comment by michelengelen** [comment]
this will show the `not contains` filter in the panel:

```jsx
export default function CustomRatingOperator() {
  const { data } = useDemoData({
    dataSet: 'Employee',
    visibleFields: VISIBLE_FIELDS,
    rowLength: 100,
  });

  const columns = React.useMemo(
    () =>
      data.columns.map((col) => {
        if (col.field === 'rating') {
          return {
            ...col,
            filterOperators: ratingOnlyOperators,
          };
        }
        if (col.field === 'name') {
          return {
            ...col,
            filterOperators: stringOperator,
          };
        }
        return col;
      }),
    [data.columns],
  );

  return (
    <div style={{ height: 400, width: '100%' }}>
      <DataGrid
        {...data}
        columns={columns}
        localeText={{
          filterOperatorNoContain: 'not contains',
        }}
      />
    </div>
  );
}
```

your error was with the `useMemo` you passed 3 arguments to it, so it got never initialised correctly.

--

One piece of advice:

I've noticed that sometimes the code you share in your issue reports can be a bit hard to follow. Here's a quick tip that might be super helpful: Following best practices for React and general coding principles, like using linters and maintaining a clean code structure, can often help catch issues early on in the development process. This means you might be able to solve some problems yourself before even posting an issue!

By following these tips, you'll be making it easier for both yourself and the community to diagnose problems and find solutions quickly. Plus, you'll gain valuable coding skills that will benefit you in the long run!

**Comment by lisaWriteJava** [comment]
Hello, 

Thanks for the feedback. I'm forking several sandboxes and combining the code into 1 to assist me in building a demo for feedback. I suspect that it why things are a hard to follow.  I've tried to clean them up as best I cant. Thanks for your help I see what I did wrong in the code to cause the 'not contains' to not show.    However, in the [original code](https://codesandbox.io/p/sandbox/material-demo-forked-n9qksl) sandbox while and old version, the 'not contains worked'. I removed the initial value and it still works [here](https://codesandbox.io/p/sandbox/material-demo-forked-n9qksl). However the [modified code](https://stackblitz.com/edit/react-nqs1xg-un48pi?file=Demo.js) which reflects a newer version of MUI  does show the not contains but it doesn't work while the original version does.. I don't get what could be wrong.

**Comment by michelengelen** [comment]
> Hello,
> 
> Thanks for the feedback. I'm forking several sandboxes and combining the code into 1 to assist me in building a demo for feedback. I suspect that it why things are a hard to follow. I've tried to clean them up as best I cant. Thanks for your help I see what I did wrong in the code to cause the 'not contains' to not show. However, in the [original code](https://codesandbox.io/p/sandbox/material-demo-forked-n9qksl) sandbox while and old version, the 'not contains worked'. I removed the initial value and it still works [here](https://codesandbox.io/p/sandbox/material-demo-forked-n9qksl). However the [modified code](https://stackblitz.com/edit/react-nqs1xg-un48pi?file=Demo.js) which reflects a newer version of MUI does show the not contains but it doesn't work while the original version does.. I don't get what could be wrong.

Hey there! No worries, I'm happy to jump in again. Here's the updated example with the "not contains" filter working on the names column: [link to updated example](https://codesandbox.io/p/sandbox/vigilant-worker-cxzgy3).

Looks like the `getApplyFilterFn` function signature changed a bit recently. It happens sometimes with library updates (like `v6` to `v7`).  Just a heads up for the future, double-checking the function signatures in the docs can save some time!

To empower you to tackle similar issues yourself, consider using StackOverflow or our awesome [community Discord server](https://mui.com/r/discord/) for debugging and code-fixing questions. The folks there are fantastic at helping with coding challenges!

**Comment by lisaWriteJava** [comment]
Thanks for your help.  Just to confirm, when I wish add this custom filter I need to  create and apply it to each column specifically as i've done [here](https://stackblitz.com/edit/react-nqs1xg-un48pi?file=Demo.js)  to add country. That There isn't away to create a custom filter than can be used for all our multiple columns. For instance the contains is there for each table for  the string columns. 

Also what did I do wrong in the example to cause it not working.  

thanks for the heads up. In the coming weeks I'll be updating the code base to v7. I'll keep an eye out for it. Right now the valueFormatter changes has my hands full.

**Comment by lisaWriteJava** [comment]
Thanks this looks good but where is: getNotContainsFilter

**Comment by lisaWriteJava** [comment]
Everything looks good, Thanks,
